linkerd/app/outbound/src/logical.rs

```diff
@@ -134,34 +143,66 @@ impl<P> svc::Param<ConcreteAddr> for Concrete<P> {
     }
 }
 
+// === impl UnwrapLogical ===
+
+fn unwrap_logical<T, P>(
+    (profile, target): (Option<profiles::Receiver>, T),
+) -> Result<svc::Either<Logical<P>, T>, Error>
+where
+    Logical<P>: From<(LogicalAddr, profiles::Receiver, T)>,
+{
+    let profile = match profile {
+        Some(profile) => profile,
+        None => {
+            debug!("No profile resolved for this target");
+            return Ok(svc::Either::B(target));
+        }
+    };
+    let addr = profile.borrow().addr.clone();
+    Ok(match addr {
+        Some(logical_addr) => svc::Either::A(Logical::from((logical_addr, profile, target))),
+        None => {
+            debug!(profile = ?*profile.borrow(), "No logical address for this profile");
+            svc::Either::B(target)
+        }
+    })
+}
+
 // === impl Outbound ===
 
 impl<L> Outbound<L> {
     /// Pushes a layer that unwraps the [`Logical`] address of a given target
     /// from its profile resolution, or else falls back to the provided
     /// per-endpoint service if there was no profile resolution for that target.
-    pub fn push_unwrap_logical<T, E>(
+    pub fn push_unwrap_logical<T, E, R, ESvc, LSvc, P>(
         self,
         endpoint: E,
     ) -> Outbound<
-        impl svc::NewService<
-                (Option<profiles::Receiver>, T),
-                Service = UnwrapLogical<L::Service, E::Service>,
-            > + Clone,
+        impl svc::NewService<(Option<profiles::Receiver>, T), Service = UnwrapLogical<LSvc, ESvc>>
+            + Clone,
     >
     where
-        tcp::Logical: From<(profiles::Receiver, T)>,
-        L: svc::NewService<tcp::Logical> + Clone,
-        E: svc::NewService<T> + Clone,
+        Logical<P>: From<(LogicalAddr, profiles::Receiver, T)>,
+        L: svc::NewService<Logical<P>, Service = LSvc> + Clone,
+        LSvc: svc::Service<R, Error = Error>,
+        LSvc::Future: Send,
+        E: svc::NewService<T, Service = ESvc> + Clone,
+        ESvc: svc::Service<R, Response = LSvc::Response, Error = Error>,
+        ESvc::Future: Send,
     {
         let Self {
             config,
             runtime,
             stack: logical,
         } = self;
         let stack = logical
-            .push_map_target(tcp::Logical::from)
-            .push(svc::UnwrapOr::layer(endpoint));
+            .push(svc::layer::mk(move |primary| {
+                svc::Filter::new(
+                    stack::NewEither::new(primary, endpoint.clone()),
+                    unwrap_logical,
+                )
+            }))
```

Comment:
Is this different than `push_switch`? This looks like a `push_switch`.

I'd generally prefer to keep the predicate inlined so the logic for consistency.